File: Jasmine.FastenerDepartment.NET/Jasmine.FastenerDepartment.Domain/Products/Models/ProductChangeReasonCode.cs
namespace Jasmine.FastenerDepartment.Domain.Products.Models;

/// <summary>
/// Product change reason code.
/// </summary>
public enum ProductChangeReasonCode
{
    /// <summary>
    /// Product has been created.
    /// </summary>
    Created = 1,

    /// <summary>
    /// Product number has been changed.
    /// </summary>
    ChangedNumber = 2,

    /// <summary>
    /// Product name has been changed.
    /// </summary>
    ChangedName = 3,

    /// <summary>
    /// Product price has been changed.
    /// </summary>
    ChangedPrice = 4,

    /// <summary>
    /// Product price tag code has been changed.
    /// </summary>
    ChangedPriceTagCode = 5,

    /// <summary>
    /// Product measurement unit code has been changed.
    /// </summary>
    ChangedMeasurementUnitCode = 6,

    /// <summary>
    /// Product order status has been changed.
    /// </summary>
    ChangedOrderStatus = 7,

    /// <summary>
    /// Product print status has been changed.
    /// </summary>
    ChangedPrintStatus = 8,

    /// <summary>
    /// Product has been deleted.
    /// </summary>
    Deleted = 9,

    /// <summary>
    /// Product has been recovered.
    /// </summary>
    Recovered = 10,

    /// <summary>
    /// Product type has been changed.
    /// </summary>
    ChangedType = 11
}

File: JasmineFastenerDepartment.Kotlin/app/src/main/java/com/szaytsev/jasminefastenerdepartment/data/models/ProductChangeReasonCode.kt
package com.szaytsev.jasminefastenerdepartment.data.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class ProductChangeReasonCode {

    @SerialName(value = "0")
    Unknown,

    @SerialName(value = "1")
    Created,

    @SerialName(value = "2")
    ChangedNumber,

    @SerialName(value = "3")
    ChangedName,

    @SerialName(value = "4")
    ChangedPrice,

    @SerialName(value = "5")
    ChangedPriceTagCode,

    @SerialName(value = "6")
    ChangedMeasurementUnitCode,

    @SerialName(value = "7")
    ChangedIsNeededToOrder,

    @SerialName(value = "8")
    ChangedIsNeededToPrint,

    @SerialName(value = "9")
    Deleted,

    @SerialName(value = "10")
    Recovered,

    @SerialName(value = "11")
    ChangedType;

    companion object {
        fun fromString(value: String) =
            ProductChangeReasonCode.entries.firstOrNull { it.ordinal.toString() == value }
                ?: throw Exception("Unexcepted change reason - $value")
    }
}